akka-stream/src/main/scala/akka/stream/impl/io/SynchronousFilePublisher.scala

```diff
@@ -0,0 +1,125 @@
+/**
+ * Copyright (C) 2015 Typesafe Inc. <http://www.typesafe.com>
+ */
+package akka.stream.impl.io
+
+import java.io.{ File, RandomAccessFile }
+import java.nio.channels.FileChannel
+
+import akka.actor.{ ActorLogging, DeadLetterSuppression, Props }
+import akka.io.DirectByteBufferPool
+import akka.stream.actor.ActorPublisherMessage
+import akka.util.ByteString
+
+import scala.annotation.tailrec
+import scala.concurrent.Promise
+
+/** INTERNAL API */
+private[akka] object SynchronousFilePublisher {
+  def props(f: File, completionPromise: Promise[Long], chunkSize: Int, initialBuffer: Int, maxBuffer: Int) = {
+    require(chunkSize > 0, s"chunkSize must be > 0 (was $chunkSize)")
+    require(initialBuffer > 0, s"initialBuffer must be > 0 (was $initialBuffer)")
+    require(maxBuffer >= initialBuffer, s"maxBuffer must be >= initialBuffer (was $maxBuffer)")
+
+    Props(classOf[SynchronousFilePublisher], f, completionPromise, chunkSize, initialBuffer, maxBuffer)
+  }
+
+  final case object Continue extends DeadLetterSuppression
+}
+
+/** INTERNAL API */
+private[akka] class SynchronousFilePublisher(f: File, bytesReadPromise: Promise[Long], chunkSize: Int, initialBuffer: Int, maxBuffer: Int)
+  extends akka.stream.actor.ActorPublisher[ByteString]
+  with ActorLogging {
+
+  import SynchronousFilePublisher._
+
+  val buffs = new DirectByteBufferPool(chunkSize, maxBuffer)
+  var eofReachedAtOffset = Long.MinValue
+
+  var readBytesTotal = 0L
+  var availableChunks: Vector[ByteString] = Vector.empty
```

Comment:
It is allowed, but usually not wise :) It will block the encapsulating actor. If you fuse multiple stages into one actor, then one blocking call would block the whole island together, which is usually fine, but can be very surprising with detached and async stages. I believe though that there are valid use cases to fuse certain stages together with a blocking one.

The intended API will require explcit denotations of fusing boundaries, so it will not be automatically done, only when you mark a composite module (graph) eligible for fusing. There might be also annotations to prevent fusing for certain stages (which would override the inherited PleaseFuse attribute from the encapsulating composite). But this is not yet really designed. We need to build the internal layout introspection/rewriting support, I would like to add a more flexible n-to-n port fusable abstraction (fusable graph stages) and then we can do it.

So for example in the case of this Source, we can set the default attribute to prohibit fusing, so people cannot accidentally put it in a larger graph and fuse it, but users will be able to override it by changing the attribute of the Source itself. This is just speculation currently, I don't have a proper design.